File: app/src/main/java/com/iku/PasswordInputActivity.java
package com.iku;

import android.app.ProgressDialog;
import android.content.Intent;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.util.Log;
import android.view.View;
import android.view.WindowManager;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.analytics.FirebaseAnalytics;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.iku.databinding.ActivityPasswordInputBinding;

import java.text.SimpleDateFormat;
import java.util.Date;

public class PasswordInputActivity extends AppCompatActivity {

    private ActivityPasswordInputBinding binding;

    private FirebaseAuth firebaseAuth;

    private FirebaseUser user;

    private FirebaseAnalytics mFirebaseAnalytics;

    private ProgressDialog mProgress;
    private static final String TAG = PasswordInputActivity.class.getSimpleName();


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityPasswordInputBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        mFirebaseAnalytics = FirebaseAnalytics.getInstance(this);

        Bundle extras = getIntent().getExtras();
        final String enteredEmail = extras.getString("email");

        firebaseAuth = FirebaseAuth.getInstance();

        getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_VISIBLE);

        initProgressDialog();

        binding.backButton.setOnClickListener(view -> onBackPressed());

        binding.signinButton.setOnClickListener(view -> {

            if (binding.enterPassword.getText().toString().isEmpty()) {
                Toast.makeText(PasswordInputActivity.this, "Please enter your password", Toast.LENGTH_SHORT).show();
                return;
            } else {
                mProgress.show();
                firebaseAuth.signInWithEmailAndPassword(enteredEmail, binding.enterPassword.getText().toString())
                        .addOnCompleteListener(PasswordInputActivity.this, task -> {
                            if (task.isSuccessful()) {
                                user = firebaseAuth.getCurrentUser();
                                FirebaseFirestore rootRef = FirebaseFirestore.getInstance();
                                DocumentReference docIdRef = rootRef.collection("users").document(user.getUid());
                                docIdRef.get().addOnCompleteListener(new OnCompleteListener<DocumentSnapshot>() {
                                    @Override
                                    public void onComplete(@NonNull Task<DocumentSnapshot> task) {
                                        if (task.isSuccessful()) {
                                            DocumentSnapshot document = task.getResult();
                                            if (document.exists()) {

                                                Toast.makeText(PasswordInputActivity.this, "Login successful", Toast.LENGTH_LONG).show();
                                                //sending to Home Activity
                                                mProgress.dismiss();
                                                Intent goToHomeActivity = new Intent(PasswordInputActivity.this, HomeActivity.class);
                                                goToHomeActivity.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
                                                startActivity(goToHomeActivity);

                                                //log event
                                                Bundle password_bundle = new Bundle();
                                                password_bundle.putString(FirebaseAnalytics.Param.METHOD, "Email");
                                                mFirebaseAnalytics.logEvent(FirebaseAnalytics.Event.LOGIN, password_bundle);
                                            } else {
                                                Toast.makeText(PasswordInputActivity.this, "User profile not filled!", Toast.LENGTH_LONG).show();
                                                //sending to NameInput Activity
                                                mProgress.dismiss();
                                                Intent goToNameInputActivity = new Intent(PasswordInputActivity.this, NameInputActivity.class);
                                                goToNameInputActivity.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
                                                startActivity(goToNameInputActivity);
                                            }
                                        } else {
                                            Log.d(TAG, "Failed with: ", task.getException());
                                        }
                                    }
                                });


                            } else {
                                mProgress.dismiss();

                                //log event
                                Bundle fail_bundle = new Bundle();
                                fail_bundle.putString(FirebaseAnalytics.Param.METHOD, "Email");
                                fail_bundle.putString("failure_reason", "incorrect password");
                                mFirebaseAnalytics.logEvent(FirebaseAnalytics.Event.LOGIN, fail_bundle);
                                Toast.makeText(PasswordInputActivity.this, "Incorrect password", Toast.LENGTH_LONG).show();
                            }
                        });
            }
        });

        binding.forgotPasswordTextView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                firebaseAuth.sendPasswordResetEmail(enteredEmail)

                        .addOnSuccessListener(aVoid -> {
                            Toast.makeText(PasswordInputActivity.this, "Password reset instructions sent via email", Toast.LENGTH_LONG).show();
                            //log event
                            Bundle passwordReset_bundle = new Bundle();
                            passwordReset_bundle.putString("state", "sent");
                            mFirebaseAnalytics.logEvent("password_reset_email", passwordReset_bundle);
                        })
                        .addOnFailureListener(e -> {
                            Toast.makeText(PasswordInputActivity.this, "Email Not Sent" + e.getMessage(), Toast.LENGTH_LONG).show();
                            //log event
                            Bundle failed_bundle = new Bundle();
                            failed_bundle.putString("state", "failed");
                            mFirebaseAnalytics.logEvent("password_reset_email", failed_bundle);

                        });
                new CountDownTimer(60000, 1000) {

                    public void onTick(long millisUntilFinished) {
                        binding.forgotPasswordTextView.setEnabled(false);
                        binding.forgotPasswordTextView.setText("Resend in " + new SimpleDateFormat("ss").format(new Date(millisUntilFinished)) + "s");
                    }

                    public void onFinish() {
                        binding.forgotPasswordTextView.setText("Forgot Password?");
                        binding.forgotPasswordTextView.setEnabled(true);

                    }
                }.start();
            }

        });

//        binding.forgotPasswordTextView.setOnClickListener(view ->
//                firebaseAuth.sendPasswordResetEmail(enteredEmail).addOnSuccessListener(
//                        aVoid ->
//                Toast.makeText(PasswordInputActivity.this, "Password reset instructions sent via email", Toast.LENGTH_LONG).show()).addOnFailureListener(e ->
//                Toast.makeText(PasswordInputActivity.this, "Email Not Sent" + e.getMessage(), Toast.LENGTH_LONG).show()));

    }

    private void initProgressDialog() {
        mProgress = new ProgressDialog(this);
        mProgress.setCancelable(false);
        mProgress.setIndeterminate(true);
    }
}

File: app/src/main/java/com/iku/PasswordInputActivity.kt
package com.iku

import android.content.Intent
import android.os.Bundle
import android.os.CountDownTimer
import android.view.View
import android.view.WindowManager
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.widget.doOnTextChanged
import com.google.android.gms.tasks.Task
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.analytics.ktx.analytics
import com.google.firebase.analytics.ktx.logEvent
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.iid.FirebaseInstanceId
import com.google.firebase.ktx.Firebase
import com.iku.app.AppConfig
import com.iku.databinding.ActivityPasswordInputBinding
import java.text.SimpleDateFormat
import java.util.*


class PasswordInputActivity : AppCompatActivity() {
    private lateinit var binding: ActivityPasswordInputBinding
    private lateinit var mAuth: FirebaseAuth
    private lateinit var mFirebaseAnalytics: FirebaseAnalytics
    private lateinit var db: FirebaseFirestore
    private lateinit var user: FirebaseUser
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityPasswordInputBinding.inflate(layoutInflater)
        setContentView(binding.root)
        mAuth = FirebaseAuth.getInstance()
        db = FirebaseFirestore.getInstance()
        mFirebaseAnalytics = Firebase.analytics
        val email = intent.getStringExtra("email")
        window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_VISIBLE)
        binding.enterPassword.doOnTextChanged { _, _, _, _ -> binding.enterPasswordLayout.isErrorEnabled = false }
        binding.enterPassword.requestFocus()
        binding.backButton.setOnClickListener { onBackPressed() }
        binding.signinButton.setOnClickListener {
            progressBarView(true)
            binding.signinButton.isEnabled = false
            if (binding.enterPassword.text.toString().isEmpty()) {
                progressBarView(false)
                binding.signinButton.isEnabled = true
                Toast.makeText(this, "Please enter your password.", Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            } else {
                mAuth.signInWithEmailAndPassword(email.toString(), binding.enterPassword.text.toString()).addOnCompleteListener { tasks: Task<AuthResult?> ->
                    if (tasks.isSuccessful) {
                        FirebaseInstanceId.getInstance().instanceId.addOnCompleteListener {
                            if (it.isSuccessful) {
                                // Token
                                val token = it.result.token
                                user = mAuth.currentUser!!
                                db.collection(AppConfig.USERS_COLLECTION).document(user.uid).get().addOnCompleteListener { task: Task<DocumentSnapshot> ->
                                    if (task.isSuccessful) {
                                        if (task.result.exists()) {
                                            sendRegistrationToken(token, user.uid)
                                            mFirebaseAnalytics.logEvent(FirebaseAnalytics.Event.LOGIN) {
                                                param(FirebaseAnalytics.Param.METHOD, "Email")
                                                param("status", "login_successful")
                                            }
                                            startActivity(Intent(this, HomeActivity::class.java).apply { addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK) })
                                        } else {
                                            mFirebaseAnalytics.logEvent(FirebaseAnalytics.Event.LOGIN) {
                                                param(FirebaseAnalytics.Param.METHOD, "Email")
                                                param("user_returned", "completing_name_input")
                                            }
                                            startActivity(Intent(this, NameInputActivity::class.java).apply { addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK) })
                                        }
                                    } else {
                                        binding.signinButton.isEnabled = true
                                    }
                                }
                            }
                        }
                    } else {
                        progressBarView(false)
                        binding.signinButton.isEnabled = true
                        mFirebaseAnalytics.logEvent(FirebaseAnalytics.Event.LOGIN) {
                            param(FirebaseAnalytics.Param.METHOD, "Email")
                            param("failure_reason", "incorrect_password")
                        }
                        binding.enterPasswordLayout.error = "Incorrect Password!"
                    }
                }
            }
        }
        binding.forgotPasswordTextView.setOnClickListener {
            mAuth.sendPasswordResetEmail(email.toString()).addOnSuccessListener {
                Toast.makeText(this, "Password reset instructions sent via email", Toast.LENGTH_LONG).show()
                mFirebaseAnalytics.logEvent("password_reset_email") {
                    param("state", "password_reset_sent")
                }
            }.addOnFailureListener {
                Toast.makeText(this, "Email not sent. Contact support.", Toast.LENGTH_LONG).show()
                mFirebaseAnalytics.logEvent("password_reset_email") {
                    param("state", "Failed_for " + user.uid)
                }
            }
            object : CountDownTimer(60000, 1000) {
                override fun onTick(millisUntilFinished: Long) {
                    binding.forgotPasswordTextView.isEnabled = false
                    binding.forgotPasswordTextView.text = applicationContext.getString(R.string.resend_email_timer_text, SimpleDateFormat("ss", Locale.US).format(Date(millisUntilFinished)))
                }

                override fun onFinish() {
                    binding.forgotPasswordTextView.isEnabled = true
                    binding.forgotPasswordTextView.text = getString(R.string.forgot_password_question)
                }
            }.start()
        }
    }

    override fun onResume() {
        binding.signinButton.isEnabled = true
        super.onResume()
    }

    private fun progressBarView(value: Boolean) {
        if (value) {
            binding.progressBar.visibility = View.VISIBLE
            window.setFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE, WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE)
        } else {
            binding.progressBar.visibility = View.GONE
            window.clearFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE)
        }
    }

    private fun sendRegistrationToken(token: String, uid: String) {
        val data = mapOf(
                "registrationToken" to token,
                "uid" to uid
        )
        db.collection(AppConfig.REGISTRATION_TOKENS_COLLECTION).document(uid).set(data).addOnSuccessListener { }.addOnFailureListener { }
    }
}